storage/zynq7000-flash/flashdrv.c

```diff
@@ -398,126 +357,424 @@ ssize_t flashdrv_read(addr_t offs, void *buff, size_t len, time_t timeout)
+		freeSz = sectSz - fdrv_common.regs[regID].buffPos;
+		chunkSz = (freeSz > (size - saveSz)) ? (size - saveSz) : freeSz;
+		memcpy(fdrv_common.regs[regID].buff + fdrv_common.regs[regID].buffPos, (const uint8_t *)data + saveSz, chunkSz);
 
 		saveSz += chunkSz;
-		fdrv_common.buffPos += chunkSz;
+		fdrv_common.regs[regID].buffPos += chunkSz;
 
-		if (fdrv_common.buffPos < sectSz)
+		if (fdrv_common.regs[regID].buffPos < sectSz) {
 			continue;
+		}
 
-		res = flashdrv_sync();
-		if (res < 0)
+		res = _flashdrv_sync(regID, strg->start);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[regID].lock);
 			return res;
+		}
 	}
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
 
 	return saveSz;
 }
 
 
-int flashdrv_done(void)
+static int flashdrv_blkSync(struct _storage_t *strg)
+{
+	int res;
+
+	if (strg == NULL || strg->dev == NULL) {
+		return -EINVAL;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	res = _flashdrv_sync(strg->dev->id, strg->start);
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return res;
+}
+
+
+const static storage_blkops_t blkOps = {
+	.read = flashdrv_blkRead,
+	.write = flashdrv_blkWrite,
+	.sync = flashdrv_blkSync
+};
+
+
+/* MTD interface */
+
+static int flashdrv_mtdErase(struct _storage_t *strg, off_t offs, size_t size)
 {
-	int res = EOK;
+	int res;
+	size_t len = 0;
+	size_t secSz = CFI_SIZE_SECTION(fdrv_common.info.cfi.regs[strg->dev->id].size);
+
+	if ((offs == 0) && (size == CFI_SIZE_FLASH(fdrv_common.info.cfi.chipSize))) {
+		mutexLock(fdrv_common.regs[strg->dev->id].lock);
+		res = _flashdrv_chipErase(strg->dev->id);
+		mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+		return res;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	while (len < size) {
+		res = _flashdrv_sectorErase(strg->dev->id, offs + strg->start + len);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+			return res;
+		}
+
+		len += secSz;
+	}
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return EOK;
+}
+
+
+static ssize_t flashdrv_mtdRead(struct _storage_t *strg, off_t offs, void *data, size_t len)
+{
+	int res;
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	res = _flashdrv_read(strg->dev->id, offs + strg->start, data, len, TIMEOUT_CMD_MS * len);
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return res;
+}
+
+
+static ssize_t flashdrv_mtdWrite(struct _storage_t *strg, off_t offs, const void *data, size_t len)
+{
+	ssize_t res = 0;
+	size_t tempSz = 0, chunkSz = 0;
+
+	chunkSz = offs % strg->dev->mtd->writeBuffsz;
+	if (chunkSz != 0) {
+		chunkSz = len < (strg->dev->mtd->writeBuffsz - chunkSz) ? len : (strg->dev->mtd->writeBuffsz - chunkSz);
+		mutexLock(fdrv_common.regs[strg->dev->id].lock);
+		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs, data, chunkSz);
+		mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+		if (res < 0) {
+			return res;
+		}
+
+		tempSz = res;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
```

Comment:
Less pretty, but we don't give lock away between operations 
```suggestion
	if (chunkSz != 0) {
		chunkSz = len < (strg->dev->mtd->writeBuffsz - chunkSz) ? len : (strg->dev->mtd->writeBuffsz - chunkSz);
		mutexLock(fdrv_common.regs[strg->dev->id].lock);
		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs, data, chunkSz);
		if (res < 0) {
			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
			return res;
		}

		tempSz = res;
	}
	else {
		mutexLock(fdrv_common.regs[strg->dev->id].lock);
	}
```

Reply:
Or simpler solution
```suggestion
	mutexLock(fdrv_common.regs[strg->dev->id].lock);
	if (chunkSz != 0) {
		chunkSz = len < (strg->dev->mtd->writeBuffsz - chunkSz) ? len : (strg->dev->mtd->writeBuffsz - chunkSz);
		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs, data, chunkSz);
		if (res < 0) {
			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
			return res;
		}

		tempSz = res;
	}
```